File: ghidra/scripts/DumpScalars.py
# Print exact little-endian scalar values from named addresses.
# Usage with analyzeHeadless:
#   -postScript DumpScalars.py f32:0066c958 f32:0066d400
#@category OpenRebellion

from java.lang import Double, Float


for spec in getScriptArgs():
    kind, raw_address = spec.split(":", 1)
    address = toAddr(raw_address)
    if kind == "f64":
        signed_long = getLong(address)
        bits = signed_long & 0xffffffffffffffff
        value = Double.longBitsToDouble(signed_long)
        println("%s bits=0x%016x value=%s" % (spec, bits, value))
        continue

    signed_bits = getInt(address)
    bits = signed_bits & 0xffffffff
    if kind == "f32":
        value = Float.intBitsToFloat(signed_bits)
    elif kind == "u32":
        value = bits
    elif kind == "i32":
        value = bits if bits < 0x80000000 else bits - 0x100000000
    else:
        printerr("unsupported scalar kind: %s" % kind)
        continue
    println("%s bits=0x%08x value=%s" % (spec, bits, value))

File: ghidra/scripts/DumpScalars.java
// Print exact little-endian scalar values from named addresses.
// Usage with analyzeHeadless:
//   -postScript DumpScalars.java f32:0066c958 f32:0066d400
//@category OpenRebellion

import ghidra.app.script.GhidraScript;

public class DumpScalars extends GhidraScript {
    @Override
    public void run() throws Exception {
        for (String spec : getScriptArgs()) {
            String[] parts = spec.split(":", 2);
            if (parts.length != 2) {
                printerr("invalid scalar specification: " + spec);
                continue;
            }

            String kind = parts[0];
            var address = toAddr(parts[1]);
            if ("f64".equals(kind)) {
                long bits = getLong(address);
                println(String.format(
                    "%s bits=0x%016x value=%s",
                    spec,
                    bits,
                    Double.longBitsToDouble(bits)
                ));
                continue;
            }

            int signedBits = getInt(address);
            long bits = Integer.toUnsignedLong(signedBits);
            String value;
            switch (kind) {
                case "f32":
                    value = Float.toString(Float.intBitsToFloat(signedBits));
                    break;
                case "u32":
                    value = Long.toString(bits);
                    break;
                case "i32":
                    value = Integer.toString(signedBits);
                    break;
                default:
                    printerr("unsupported scalar kind: " + kind);
                    continue;
            }
            println(String.format("%s bits=0x%08x value=%s", spec, bits, value));
        }
    }
}
